fix garmin serial/usb device character encoding/decoding.

tsteven4 opened the pull request:
We got away with breaking this for a long time, but we finally got caught.

This resolves #1115

@svintuss has tested both read and write on his Garmin GPSmap 60cx using windows-1251.

tsteven4 committed:
attempt to use correct codec with garmin reader.

the garmin writer is unchanged as of yet.

tsteven4 committed:
fix garmin writer wrt encoding.

tsteven4 committed:
add garmin option to force codec.

tsteven4 committed:
Merge branch 'master' into garmincodec

tsteven4 committed:
update serialization ref files for garmin codec option.

tsteven4 committed:
fix potential overwrite bug with route/track names

in garmin writer.

tsteven4 committed:
fix up for old Qt.

tsteven4 committed:
correct assertion

tsteven4 committed:
use static_assert to for fixed buffer size checks.

robertlipe commented on garmin.cc:
```diff
@@ -347,7 +368,26 @@ rw_init(const QString& fname)
+  assert(!QString(valid_waypt_chars).contains('^'));
+  assert(!QString(valid_waypt_chars).contains('-'));
+  assert(!QString(valid_waypt_chars).contains('['));
+  assert(!QString(valid_waypt_chars).contains(']'));
```
These could be a single QString::contains(const [QRegularExpression](https://doc.qt.io/qt-6/qregularexpression.html) &re, but it would admittedly be a quoting nightmare and wouldn't give you a single error message like this will. Your call.

tsteven4 replied:
Yes, and it would save a bunch of QString construction.  But, the code is only run on init, and only in debug builds, so I thought clarity was better.

robertlipe commented on garmin.cc:
```diff
@@ -347,7 +368,26 @@ rw_init(const QString& fname)
    */
+  if (opt_codec != nullptr) {
+    // override expected codec with user supplied choice.
+    receiver_charset = opt_codec;
```
This is a good option - though we should set low expectations for it in the doc - but if he has an ID 292 receiver, why does he need to specify this option?

Or was that just for testing? I'll admit I didn't follow the play-by-play.

tsteven4 replied:
Because he is using windows-1251 to get Cyrillic characters instead of the default windows-1252 Latin.

robertlipe commented on garmin.cc:
```diff
@@ -347,7 +368,26 @@ rw_init(const QString& fname)
+  assert(!QString(valid_waypt_chars).contains('['));
+  assert(!QString(valid_waypt_chars).contains(']'));
+  invalid_char_re = QRegularExpression(QStringLiteral("[^%1]").arg(valid_waypt_chars));
+  assert(invalid_char_re.isValid());
```
I'll grant we should never ever see these asserts failing, but triggering a failing assert inside rw_init - or after - can leave the device in a dangling state that can require power cycling the receiver to recover from. This same issue is shared with the calls to fatal(), so I won't argue with it strongly.

This is one of those rare textbook cases where exceptions might actually make sense. Our own _deinit() doesn't really do any meaningful cleanup, for much the same reason dtors aren't called on failed ctors.

Then again, we're not flagrantly in the middle of a _Write_Packet/Get_Ack cycle (which I KNOW will hang many devices).

Meh. It's an exceptional case and I don't really know that we can do anything else better.

No action required, but this may be educational, so I'm not deleting it...

tsteven4 replied:
Its kind of tacky to report a code problem as a warning, but that would avoid leaving the device dangling.

robertlipe replied:
I *think* this is early enough that it won't, in practice, hang. Since the triggered case would be 100% here, hanging the device or not isn't a big deal if it even DOES hang the device.

robertlipe commented on garmin.cc:
```diff
@@ -418,8 +458,8 @@ waypt_read()
 
-    wpt_tmp->shortname = QString::fromLatin1(way[i]->ident);
-    wpt_tmp->description = QString::fromLatin1(way[i]->cmnt);
+    wpt_tmp->shortname = str_to_unicode(way[i]->ident);
```
Should we give those foo[NNN]'s in jeeps/gps.h a better type than 'char' so we could add some meaningful overrides here and not live in constant fear of failing to zero pad them and/or character convert them?

Optional.

tsteven4 replied:
If we were starting over certainly. We do treat some as zero padded and require zero termination, and others (in fs data) as zero padded but not necessarily zero terminated.  To me it isn't really clear what the requirements are, or even if the requirements were constant across all the devices.  The fs data ones generate Wstringop-truncation warnings.  But I'm not messing with jeeps today.

robertlipe replied:
Good call.
The original JEEPS API used zero-padded buffers everywhere and we had all kinds of fits when initially using that with C strings. It's very likely that the newer code (e.g. D108, D109) uses C strings instead of zero-padded ones like D103 originally would have used. It's likely currently a mixture and that's indeed messy.
If we were starting over with jeeps, it would be without jeeps and the Garmin protocol implementations would look very very different. Running away is a a good decision. The payoff in picking at this scab isn't worth it.

robertlipe commented on reference/help.txt:
```diff
@@ -137,6 +137,7 @@ File Types (-i and -o options):
 	  category              Category number to use for written waypoints
 	  bitscategory          Bitmap of categories
 	  baud                  Speed in bits per second of serial port (baud=9600
+	  codec                 override codec to use for device
```
Uppercase, please.
Needs a corresponding option_codec.xml in docbook.

tsteven4 replied:
done

robertlipe commented on garmin.cc:
```diff
@@ -134,8 +142,22 @@ static int d103_icon_number_from_symbol(const QString& s);
+}
+
+static void
+write_char_string(char* dest, const char* source, size_t destsize)
```
Good idea. I should have done that 22 years ago. :-)
Should it have a sibling that does the character set conversion from the source?
I don't know if you got them all, but almost every occurrence of CSTRc in this source file is a candidate for replacement thusly, especially those as arguments to memcpy or strncpy.

tsteven4 replied:
If we assume the source is zero terminated, which we have, then the sibling isn't necessary.

robertlipe replied:
It would have just been a single place to wrap the toLatin/fromUtf style converters. That's all.

robertlipe reviewed (approved):
I quite like it. Merge it down whenever you're ready.

tsteven4 committed:
document garmin codec option

tsteven4 committed:
don't clean rtept/trkpt names for newer devices.

We clear badchars for newer devices already, effectively not
cleaning wpt names.

tsteven4 committed:
Merge branch 'master' into garmincodec
